### ISSUE TITLE
Distinguish between F821 "undefined name" for undefined variables VS undefined string types

### ISSUE THREAD

**Issue by BenQuigley** [astral-sh/ruff#8700]
I'm running into a similar issue as https://github.com/astral-sh/ruff/issues/7175 and I see that the user who raised the issue and maintainers ended up agreeing that ruff should complain about, e.g. `def foo(bar: "Bar")` if `Bar` is not defined as a type. That makes sense to me, but my team and I would disable that particular error from ruff (which is error `F821`). We use Python type annotations mostly the same way we use code comments; that is, we view them as helpful but we don't programmatically check them, a system that works well for us.

The problem is that running ruff with `ignore=[F821]` against our code base would also disable checking for undefined variables, which are important to check for.

Example script:

```python
def foo_function(foo: "Foo"):  # type annotation crimes, but valid python
    pass

Foo  # NameError
```

ruff output:
```shell
ruff test.py
test.py:1:24: F821 Undefined name `Foo`
test.py:4:1: F821 Undefined name `Foo`
Found 2 errors.
```

I would like to be able to update my `pyproject.toml` to ignore the undefined type annotation, but not the programming error in line 4; so I would like to request for the "undefined forward type annotation" error to be broken out into a distinct error code.

**Comment by BenQuigley** [comment]
Updated the example to make a little more sense :)

**Comment by zayenzl** [comment]
I've seen this same problem, and I think it would be _very_ valuable to separate these two very different cases. Especially as [PEP 484](https://peps.python.org/pep-0484/#forward-references) explicitly says that strings can be used for forward references
> When a type hint contains names that have not been defined yet, that definition may be expressed as a string literal, to be resolved later.

**Comment by HWiese1980** [comment]
Yeah, forward referencing is a common means to mitigate circular imports and valid Python. Ruff should consider this and allow it.

**Comment by AntiSol** [comment]
It's not just circular imports, there are also other perfectly valid cases where a type might not be resolvable - that's precisely why we have the ability to place them in strings as a forward reference. 

For example, it can also be an efficiency thing - e.g I don't actually need to import all of numpy, slowing down my program's startup, just so that my configuration class can specify that a particular value is a numpy array. 

Instead I should be able to use a forward reference to signify that the type isn't resolveable - the intended use of forward references - and have ruff respect that.

This should absolutely be split out into a separate rule. I am currently in a situation where my options are:
1. Remove typing information from a couple of vars, making them harder to comprehend for everybody
2. Import numpy for a script that does not need it, slowing down script startup.
3. Disable F821, losing the advantages that rule brings when it's not spitting out false positives

**Comment by mikedh** [comment]
Thanks @BenQuigley  for raising this again with the further detail! I still have not found a good way to add type hints without adding circular imports everywhere, and I really think that forward references should be supported in more cases. Imports in Python are often expensive, and adding imports "just for the hint" seems to me to be the tail wagging the dog. 

I guess what I think the best behavior would be here is: "fully defined forward references are always allowed, i.e. `"trimesh.Trimesh"` or `"scipy.spatial.cKDTree"` can be used anywhere regardless of import status. I've just been disabling the checks, i.e. `"trimesh.Trimesh":  # type: ignore[name-defined]` but would love a better solution.
